src/backend/distributed/sql/udfs/citus_finalize_upgrade_to_citus11/latest.sql

```diff
@@ -0,0 +1,184 @@
+    -- works as intended
+    -- And, this should be done only ONCE for a cluster as it can be a pretty
+    -- time consuming operation. Thus, even if the function is called multiple time,
+    -- we keep track of it and do not re-execute this part if not needed.
+  ------------------------------------------------------------------------------------------
+  DECLARE
+      partitioned_table_exists_pre_11 boolean:=False;
+  BEGIN
+
+    -- we recorded if partitioned tables exists during upgrade to Citus 11
+    SELECT metadata->>'partitioned_citus_table_exists_pre_11' INTO partitioned_table_exists_pre_11
+    FROM pg_dist_node_metadata;
+
+    IF partitioned_table_exists_pre_11 IS NOT NULL AND partitioned_table_exists_pre_11 THEN
+
+      -- this might take long depending on the number of partitions and shards...
+      RAISE NOTICE 'Preparing all the existing partitioned table indexes';
+      PERFORM pg_catalog.fix_all_partition_shard_index_names();
+
+      -- great, we are done with fixing the existing wrong index names
+      -- so, lets remove this
+      UPDATE pg_dist_node_metadata
+      SET metadata=jsonb_delete(metadata, 'partitioned_citus_table_exists_pre_11');
+    ELSE
+        RAISE DEBUG 'There are no partitioned tables that should be fixed';
+    END IF;
+  END;
+
+  ------------------------------------------------------------------------------------------
+  -- STAGE 6: Return early if there are no primary worker nodes
+  -- We don't strictly need this step, but it gives a nicer notice message
+  ------------------------------------------------------------------------------------------
+  DECLARE
+    primary_worker_node_count bigint :=0;
+  BEGIN
+        SELECT count(*) INTO primary_worker_node_count FROM pg_dist_node WHERE groupid != 0 AND noderole = 'primary';
+
+        IF primary_worker_node_count = 0 THEN
+                  RAISE NOTICE 'There are no primary worker nodes, no need to sync metadata to any node';
+                  RETURN true;
+        ELSE
+                  RAISE DEBUG 'There are % primary worker nodes, continue to sync metadata', primary_worker_node_count;
+        END IF;
+  END;
+
+  ------------------------------------------------------------------------------------------
+  -- STAGE 7: Do the actual metadata & object syncing to the worker nodes
+  -- For the "already synced" metadata nodes, we do not strictly need to
+  -- sync the objects & metadata, but there is no harm to do it anyway
+  -- it'll only cost some execution time but makes sure that we have a
+  -- a consistent metadata & objects across all the nodes
+  ------------------------------------------------------------------------------------------
+  DECLARE
+  BEGIN
+
+    -- this might take long depending on the number of tables & objects ...
+    RAISE NOTICE 'Preparing to sync the metadata to all nodes';
+
+    PERFORM start_metadata_sync_to_node(nodename,nodeport)
```

Comment:
The one thing I wanted to is to run `start_metadata_sync_to_node` across nodes in parallel. However, that is not trivial to do. With the `all_objects.sql`, it takes ~2 minutes to complete on a 7 worker local cluster. Do you think should we spend time to parallelize this?

```SQL
select pg_catalog.citus_finalize_upgrade_to_citus11();
NOTICE:  Preparing all the existing partitioned table indexes
NOTICE:  Preparing to sync the metadata to all nodes
NOTICE:  Replicating postgres objects to node localhost:9700
DETAIL:  There are 503 objects to replicate, depending on your environment this might take a while
NOTICE:  Replicating postgres objects to node localhost:9701
DETAIL:  There are 503 objects to replicate, depending on your environment this might take a while
NOTICE:  Replicating postgres objects to node localhost:9702
DETAIL:  There are 503 objects to replicate, depending on your environment this might take a while
NOTICE:  Replicating postgres objects to node localhost:9703
DETAIL:  There are 503 objects to replicate, depending on your environment this might take a while
NOTICE:  Replicating postgres objects to node localhost:9704
DETAIL:  There are 503 objects to replicate, depending on your environment this might take a while
NOTICE:  Replicating postgres objects to node localhost:9705
DETAIL:  There are 503 objects to replicate, depending on your environment this might take a while
NOTICE:  Replicating postgres objects to node localhost:9706
DETAIL:  There are 503 objects to replicate, depending on your environment this might take a while
 citus_finalize_upgrade_to_citus11 
-----------------------------------
 t
(1 row)

Time: 131397.190 ms (02:11.397)
```